[SPARK-6602][Core] Replace direct use of Akka with Spark RPC interface - part 1

zsxwing opened the pull request:
This PR replaced the following `Actor`s to `RpcEndpoint`:
1. HeartbeatReceiver
2. ExecutorActor
3. BlockManagerMasterActor
4. BlockManagerSlaveActor
5. CoarseGrainedExecutorBackend and subclasses
6. CoarseGrainedSchedulerBackend.DriverActor

This is the first PR. I will split the work of SPARK-6602 to several PRs for code review.

zsxwing committed:
Update HeartbeatReceiver to use RpcEndpoint

zsxwing committed:
Update ExecutorActor to use RpcEndpoint

zsxwing committed:
Update BlockManagerMaster to use RpcEndpoint

zsxwing committed:
Update BlockManagerSlaveActor to use RpcEndpoint

zsxwing committed:
Update CoarseGrainedExecutorBackend and CoarseGrainedSchedulerBackend to use RpcEndpoint

zsxwing committed:
Fix some bugs after rebasing the changes on the master

SparkQA commented:
[Test build #29406 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29406/consoleFull) for   PR 5268 at commit [`705245d`](https://github.com/apache/spark/commit/705245d218ed3ec62877f7f4e43066601ac67fe9).

SparkQA commented:
[Test build #29406 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29406/consoleFull) for   PR 5268 at commit [`705245d`](https://github.com/apache/spark/commit/705245d218ed3ec62877f7f4e43066601ac67fe9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29406/
Test PASSed.

rxin commented on core/src/main/scala/org/apache/spark/storage/BlockManagerSlaveEndpoint.scala:
```diff
@@ -17,41 +17,43 @@
 
     case RemoveShuffle(shuffleId) =>
-      doAsync[Boolean]("removing shuffle " + shuffleId, sender) {
+      doAsync[Boolean]("removing shuffle " + shuffleId, context) {
```
do we need to send some reply here?

rxin replied:
oh ic it is done in doAsync. Is this ok with the RPC contract that we are replying from a different thread?

zsxwing replied:
Yes. RpcCallContext should be thread-safe and can be called in another thread. It's a very common usage. I will add docs about it.

rxin commented on core/src/main/scala/org/apache/spark/scheduler/local/LocalBackend.scala:
```diff
@@ -87,9 +86,15 @@ private[spark] class LocalActor(
       // Try to reviveOffer after 1 second, because scheduler may wait for locality timeout
-      context.system.scheduler.scheduleOnce(1000 millis, self, ReviveOffers)
+      reviveThread.schedule(new Runnable {
+        override def run(): Unit = self.send(ReviveOffers)
```
does send ever throw an exception? what happens to that thread?

rxin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -169,6 +165,11 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val actorSyste
 
+    override def onDisconnected(remoteAddress: RpcAddress): Unit = {
+      addressToExecutorId.get(remoteAddress).foreach(removeExecutor(_,
+        "remote Akka client disassociated"))
```
change the error message

zsxwing replied:
Fixed.

rxin commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -92,23 +96,28 @@ private[spark] class CoarseGrainedExecutorBackend(
+      logError(s"Driver $remoteAddress disassociated! Shutting down.")
+      System.exit(1)
+    } else {
+      logWarning(s"Received irrelevant DisassociatedEvent $remoteAddress")
```
```scala
s"An unknown ($remoteAddress) driver disconnected."
```

zsxwing replied:
Fixed.

rxin commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -21,39 +21,43 @@ import java.net.URL
+        RegisterExecutor(executorId, self, hostPort, cores, extractLogUrls))
+    } onComplete {
+      case Success(msg) => self.send(msg)
+      case Failure(e) => logError(s"Cannot register to driver: $driverUrl", e)
```
register with driver

zsxwing replied:
Fixed.

rxin commented on core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala:
```diff
@@ -21,39 +21,43 @@ import java.net.URL
 
   var executor: Executor = null
-  var driver: ActorSelection = null
+  @volatile var driver: Option[RpcEndpointRef] = None
```
how about replacing this with a AtomicReference?

rxin commented on core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala:
```diff
@@ -91,17 +103,25 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, scheduler: TaskSchedule
           s"timed out after ${now - lastSeenMs} ms"))
         if (sc.supportDynamicAllocation) {
-          sc.killExecutor(executorId)
+          // Asynchronously kill the executor to avoid blocking the current thread
```
was it not doing async before? if it wasn't, why do we need to change it here? If it was, did the RPC contract changed it so that it is no longer async?

zsxwing replied:
No.

Here is the execution of `killExecutor`:

SparkContext.killExecutor -> SparkContext.killExecutors -> CoarseGrainedSchedulerBackend.killExecutors -> CoarseGrainedSchedulerBackend.doKillExecutors

If CoarseGrainedSchedulerBackend is YarnSchedulerBackend, it will `askWithReply` KillExecutors to YarnSchedulerEndpoint, and YarnSchedulerEndpoint will `askWithReply` KillExecutors to AM. 

Because AM and YarnSchedulerEndpoint are not in the same node in yarn-client mode, it needs to go through the network and wait for a reply.

Since it's a blocking method, I need to call it asynchronously in another thread.

rxin commented:
I took a very quick look at this. @zsxwing If there are any tricky parts, can you point them out on this pull request to make it easier to review?

zsxwing committed:
Make self return null after stopping RpcEndpointRef; fix docs and error messages

SparkQA commented:
[Test build #29481 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29481/consoleFull) for   PR 5268 at commit [`30a9036`](https://github.com/apache/spark/commit/30a903617fce00e66b6ab237f8c4ec3415da5b1e).

SparkQA commented:
[Test build #29481 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/29481/consoleFull) for   PR 5268 at commit [`30a9036`](https://github.com/apache/spark/commit/30a903617fce00e66b6ab237f8c4ec3415da5b1e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.
 * This patch does not change any dependencies.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29481/
Test PASSed.

zsxwing commented:
There are two things that's worth to review carefully:

1. Calling `setupEndpointRef ` or `setupThreadSafeEndpointRef ` to create `RpcEndpoint`. When I was doing this PR, I found that it's inconvenient to check if using `setupEndpointRef ` or `setupThreadSafeEndpointRef ` correctly. The problem is I need to read the implementation of `RpcEndpoint` in one file at first, search and switch to other places that creating `RpcEndpoint`. I'm wondering if the current design of `setupEndpointRef ` and `setupThreadSafeEndpointRef` can be improved. 

One possible improvement is using a special trait, such as `ThreadSafeRpcEndpoint` to mark the RpcEndpoint that needs to be thread-safe, and only have one method, `setupEndpointRef`, to setup RpcEndpoint. We can use isInstanceOf to distinguish them.

2. If a message is passed to `send`, the receiver side should handle it in `receive`. If a message is passed to `sendWithReply` or `askWithReply`, the receiver side should handle it in `receiveAndReply`. If using a wrong method, the message will be lost (of cause, will be logged).

zsxwing commented:
> does send ever throw an exception? what happens to that thread?

`self.send` may throw an IllegalArgumentException, because both `self.send` and `onStop` may called at the same time.

I updated the contract of `self`. Now if a `RpcEndpointRef` is stopped, `self` will return null. I also added `Utils.tryLogNonFatalError`, which will catch all non-fatal errors, and use it to wrap the codes in `Runnable`s.

rxin commented:
I like the idea of ThreadSafeRpcEndpoint.

I need to think a little bit more about send/sendWithReply. I do find it error prone as is.

rxin commented:
The changes mostly lgtm. Let's add the ThreadSafeRpcEndpoint interface for differentiation.

zsxwing committed:
Rmove setupThreadSafeEndpoint and add ThreadSafeRpcEndpoint

zsxwing committed:
Merge branch 'master' into rpc-rewrite

Conflicts:
	core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala
	core/src/main/scala/org/apache/spark/SparkContext.scala
	core/src/main/scala/org/apache/spark/util/Utils.scala
	yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29710/
Test FAILed.

zsxwing committed:
Fix the code style

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/29711/
Test PASSed.

rxin commented:
Merging this in master. Thanks.
